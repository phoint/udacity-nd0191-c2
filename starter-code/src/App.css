/* NavBar */
nav {
  background-color: #f2f2f2;
  padding: 10px;
  border-bottom: 1px solid #333;
}

ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

li {
  display: inline-block;
  margin-right: 10px;
}

a {
  text-decoration: none;
  color: #333;
  padding: 5px;
}

a.active {
  border-bottom: 2px solid #333;
}

nav a:hover {
  background-color: #333;
  color: #fff;
}
/* NavBar */
.container {
  width: 80%;
  max-width: 1000px;
  margin: auto;

}

/* Dashboard */

.dashboard-container {
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 20px;
  background-color: #fff;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  margin: 10px auto;
}

h2 {
  color: #333;
  margin-bottom: 20px;
  text-align: center;
}

.questions {
  display: flex;
  flex-wrap: wrap;
  gap: 20px 10px;
}

.question-card {
  background-color: #e8f4f8;
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 20px;
  flex: 1 1 calc(25% - 20px);
  box-sizing: border-box;
  text-align: center;
}

.question-card h3 {
  margin: 0;
  color: #007bff;
}

.question-card p {
  margin: 10px 0;
  color: #666;
}

.show-button {
  background-color: #fff;
  color: #28a745;
  border: 1px solid #28a745;
  border-radius: 5px;
  padding: 10px 20px;
  cursor: pointer;
  transition: background-color 0.3s;
  width: 100%;
}

.show-button:hover {
  background-color: #218838;
  color: #fff;
}

.done {
  background-color: #f8f9fa;
  border-color: #dee2e6;
}

@media (max-width: 768px) {
  .question-card {
    flex: 1 1 calc(50% - 40px)
  }
}

@media (max-width: 480px) {
  .question-card {
    flex: 1 1 100%;
  }
}

/* Dashboard */

/* Poll Page */
.poll-container {
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 20px;
  background-color: #fff;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  margin: 10px auto;
}

h2 {
  margin-bottom: 20px;
  color: #333;
}

.poll-header {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-bottom: 30px;
}

.avatar {
  border-radius: 50%;
  width: 100px;
  height: 100px;
  margin-bottom: 10px;
}

.poll-header h3 {
  margin: 0;
  color: #007bff;
}

.poll-options {
  display: flex;
  justify-content: space-around;
}

.option {
  background-color: #e8f4f8;
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 20px;
  width: 45%;
  text-align: center;
}

.option p {
  margin-bottom: 20px;
  color: #666;
}

.vote-button {
  background-color: #28a745;
  color: #fff;
  border: none;
  border-radius: 5px;
  padding: 10px 20px;
  cursor: pointer;
  transition: background-color 0.3s;
  width: 100%;
}

.vote-button:hover {
  background-color: #218838;
}

/* Leaderboard */
.table-container {
  background-color: #fff;
  padding: 20px;
  border-radius: 10px;
  box-shadow: 0 0 15px rgba(0, 0, 0, 0.1);
  box-sizing: border-box;
}

.table-container h2 {
  margin-bottom: 20px;
  color: #333;
  text-align: center;
}

.leaderboard-table {
  width: 100%;
  border-collapse: collapse;
  margin-bottom: 20px;
}

.leaderboard-table thead {
  background-color: #007bff;
  color: #fff;
}

.leaderboard-table th,
.leaderboard-table td {
  padding: 10px;
  text-align: left;
  border-bottom: 1px solid #ddd;
}

.leaderboard-table th {
  text-align: center;
}

.leaderboard-table tbody tr:hover {
  background-color: #f1f1f1;
}

.user-block {
  display: flex;
  align-items: center;
}

.avatar {
  border-radius: 50%;
  width: 30px;
  height: 30px;
  margin-right: 10px;
}

.username {
  color: #666;
  font-size: small;
}

@media (max-width: 768px) {
  .table-container {
    width: 90%;
  }

  .leaderboard-table th,
  .leaderboard-table td {
    padding: 8px;
  }
}

/* Leaderboard */

/* New Question */
.form-container {
  background-color: #fff;
  padding: 30px;
  border-radius: 10px;
  box-shadow: 0 0 15px rgba(0, 0, 0, 0.1);
  text-align: center;
  box-sizing: border-box;
}

.sub-heading {
  color: #666;
  margin-bottom: 30px;
}

.form {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.form-group {
  width: 100%;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

label {
  margin-bottom: 5px;
  color: #333;
}

input[type="text"] {
  width: 100%;
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 5px;
  box-sizing: border-box;
}

input[type="text"]::placeholder {
  color: #ccc;
}

.submit-button {
  background-color: #007bff;
  color: #fff;
  border: none;
  border-radius: 5px;
  padding: 10px 20px;
  cursor: pointer;
  transition: background-color 0.3s;
  margin: 20px auto;
}

.submit-button:hover {
  background-color: #0056b3;
}

.submit-button:disabled {
  background-color: #bebaba;
  color: #666;
}

@media (max-width: 768px) {
  .form-container {
    width: 90%;
  }

  .form-group {
    width: 100%;;
  }
}