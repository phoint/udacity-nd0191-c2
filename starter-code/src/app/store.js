import { configureStore } from '@reduxjs/toolkit'
import logger from 'redux-logger'
import questionsReducer from '../features/questions/questionSlice'
import usersReducer from '../features/users/userSlice'
import authedUserReducer from '../features/authedUser/authedUserSlice'

export const setupStore = preloadedState => {
  return configureStore({
    reducer: {
      questions: questionsReducer,
      users: usersReducer,
      authedUser: authedUserReducer
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
    preloadedState
  })
}