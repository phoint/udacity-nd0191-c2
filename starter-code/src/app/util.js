export const LoadingStatus = Object.freeze({
    IDLE: 'idle',
    LOADING: 'loading',
    PENDING: 'pending',
    SUCCESS: 'success',
    FAILURE: 'failure'
  });