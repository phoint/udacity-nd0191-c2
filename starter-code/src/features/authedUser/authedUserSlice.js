import { createSlice } from "@reduxjs/toolkit";

export const authedUserSlice = createSlice({
    name: 'authedUser',
    initialState: null,
    reducers: {
        setAuthedUser: (state, action) => {
            state.authedUser = action.payload;
        },
        login: (state, action) => {
            state.authedUser = action.payload;
        },
        logout: (state, action) => {
            state.authedUser = null;
        }
    }
})

export const { setAuthedUser, login, logout } = authedUserSlice.actions
export default authedUserSlice.reducer

