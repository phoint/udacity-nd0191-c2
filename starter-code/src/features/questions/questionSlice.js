import { createAsyncThunk, createSelector, createSlice } from "@reduxjs/toolkit";
import { LoadingStatus } from "../../app/util";
import { _getQuestions, _saveQuestion, _saveQuestionAnswer } from "../../_DATA";

const initialState = {
    items: {},
    status: LoadingStatus.IDLE,
    error: null,
}

const questionsSlice = createSlice({
    name: 'questions',
    initialState,
    reducers: {},
    extraReducers(builder) {
        builder.addCase(fetchQuestions.pending, (state, action) => {
            state.status = LoadingStatus.PENDING
        }).addCase(fetchQuestions.fulfilled, (state, action) => {
            state.status = LoadingStatus.SUCCESS
            state.items = action.payload
        }).addCase(fetchQuestions.rejected, (state, action) => {
            state.status = LoadingStatus.FAILURE
            state.error = action.error.message
        }).addCase(addNewQuestion.pending, (state, action) => {
            state.status = LoadingStatus.PENDING
        }).addCase(addNewQuestion.fulfilled, (state, action) => {
            state.items = Object.assign({ [action.payload.id]: action.payload }, state.items)
            state.status = LoadingStatus.SUCCESS
        }).addCase(addNewQuestion.rejected, (state, action) => {
            state.status = LoadingStatus.FAILURE
            state.error = action.error.message
        }).addCase(saveQuestionAnswer.pending, (state, action) => {
            state.status = LoadingStatus.PENDING
        }).addCase(saveQuestionAnswer.fulfilled, (state, action) => {
            state.status = LoadingStatus.SUCCESS
            const { authedUser, qid, answer } = action.payload
            const questions = state.items
            const notAnswer = answer === 'optionOne' ? 'optionTwo' : 'optionOne'
            state.items = {
                ...questions,
                [qid]: {
                    ...questions[qid],
                    [answer]: {
                        ...questions[qid][answer],
                        votes: questions[qid][answer].votes.concat([authedUser])
                    },
                    [notAnswer] : {
                        ...questions[qid][notAnswer],
                        votes: questions[qid][notAnswer].votes.filter(user => user !== authedUser)
                    }
                }
            }
        })
    }
})

export const { getQuestions } = questionsSlice.actions
export default questionsSlice.reducer

export const fetchQuestions = createAsyncThunk('questions/fetchQuestions', () => _getQuestions())
export const addNewQuestion = createAsyncThunk('questions/addNewQuestion', (initialQuestion) => _saveQuestion(initialQuestion))
export const saveQuestionAnswer = createAsyncThunk('question/saveAnswer', async (answer) => {
    const isSaved = await _saveQuestionAnswer(answer)
    return {
        ...answer,
        isSaved: isSaved
    }
})

export const selectAllQuestions = createSelector([state => state.questions, (state, authedUser) => authedUser], 
    (questions, authedUser) => (
        Object.values(questions.items).map(question => ({
            id: question.id,
            author: question.author,
            timestamp: question.timestamp,
            done: question.optionOne.votes.includes(authedUser) || question.optionTwo.votes.includes(authedUser),
        })).sort((a,b) => b.timestamp - a.timestamp)
    ))

export const getQuestionById = createSelector([state => state.questions, (state, id) => id], (questions, id) => questions.items[id]);

