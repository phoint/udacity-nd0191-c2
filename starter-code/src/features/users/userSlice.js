import { createSlice, createAsyncThunk, createSelector } from "@reduxjs/toolkit";
import { _getUsers } from "../../_DATA";
import { LoadingStatus } from "../../app/util";
import { saveQuestionAnswer } from "../questions/questionSlice";

const initialState = {
    status: LoadingStatus.IDLE,
    items: {},
    error: null
}

const usersSlice = createSlice({
    name: 'user',
    initialState,
    reducers: {
    },
    extraReducers(builder) {
        builder.addCase(fetchUser.pending, (state, action) => {
            state.status = LoadingStatus.PENDING;
        }).addCase(fetchUser.fulfilled, (state, action) => {
            state.status = LoadingStatus.SUCCESS;
            state.items = action.payload;
        }).addCase(fetchUser.rejected, (state, action) => {
            state.status = LoadingStatus.FAILURE;
            state.error = action.error.message;
        }).addCase(saveQuestionAnswer.fulfilled, (state, action) => {
            state.status = LoadingStatus.SUCCESS;
            const users = state.items
            const { authedUser, qid, answer } = action.payload
            state.items = {
                ...users,
                [authedUser]: {
                    ...users[authedUser],
                    answers: {
                        ...users[authedUser].answers,
                        [qid]: answer
                    }
                }
            }
        })
    }
})

export const { getUsers } = usersSlice.actions
export default usersSlice.reducer

export const fetchUser = createAsyncThunk('users/fetchUsers', () => _getUsers())

export const getAllUsers = (state) => Object.values(state.users).map(user => ({
    id: user.id,
    name: user.name,
    avatarURL: user.avatarURL,
}))

export const selectUserByLeaderdoard = createSelector([state => state.users], 
    (users) => Object.values(users.items).map(i => {
        const {id, name, avatarURL, answers, questions} = i;
        return {
            id,
            name,
            avatarURL,
            question: questions.length,
            answer: Object.entries(answers).length
        }
    }))

export const getUserById = createSelector([state => state.users, (state, userId) => userId], (users, userId) => (users.items[userId]))
