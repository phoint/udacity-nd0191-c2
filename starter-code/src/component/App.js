import logo from '../logo.svg';
import '../App.css';
import { useDispatch, useSelector} from 'react-redux'
import { useEffect, Fragment, useRef } from 'react';
import { Dashboard } from './DashBoard';
import { fetchQuestions } from '../features/questions/questionSlice';
import { login } from '../features/authedUser/authedUserSlice';
import { fetchUser } from '../features/users/userSlice';
import { Poll } from './Poll';
import { LoadingStatus } from '../app/util';
import { NewQuestion } from './NewQuestion';
import { Leaderboard } from './Leaderboard';
import LoadingBar from 'react-top-loading-bar';
import { Route, Routes } from 'react-router-dom';
import Nav from './Nav';
import Login from './Login';
import NotFound from './NotFound';
import Protected from './Protected';

function App() {
  const dispatch = useDispatch()
  const loadingQuestion = useSelector(state => state.questions.status)
  const loadingUser = useSelector(state => state.users.status)
  const authedUser = useSelector(state => state.authedUser.id)
  const ref = useRef(null)
  useEffect(() => {
    ref.current.continuousStart();
    handleInitialData(dispatch);
  }
  ,[dispatch])

  const handleInitialData = (dispatch) => {
      // dispatch(login({
      //   id:"sarahedo",
      //   name:"Sarah Edo"
      // }));
      dispatch(fetchQuestions());
      dispatch(fetchUser());
      ref.current.complete();
  }

  return (
    <div className='container'>
      <Fragment>
        <LoadingBar color="#f11946" ref={ref} shadow="true" />
        {loadingUser === LoadingStatus.SUCCESS && <Nav />}
        <Routes>
          <Route path="/" exact element={
            <Protected>
              <Dashboard />
            </Protected>
          } />
          <Route path="/question/:id" exact element={
            <Protected>
              <Poll />
            </Protected>
          } />
          <Route path="/new" exact element={
            <Protected>
              <NewQuestion />
            </Protected>
          } />
          <Route path="/leaderboard" exact element={
            <Protected>
              <Leaderboard />
            </Protected>
          } />
          <Route path="*" exact element={
            <Protected>
              <NotFound />
            </Protected>
          } />
          <Route path='/login' exact element={<Login />} />
        </Routes>
        {/* {loadingQuestion === LoadingStatus.SUCCESS && <Dashboard/>} */}
        {/* {loadingUser === LoadingStatus.SUCCESS && <Poll id="8xf0y6ziyjabvozdd253nd"/>} */}
        {/* { authedUser && <NewQuestion/>} */}
        {/* {loadingUser === LoadingStatus.SUCCESS && <Leaderboard/>} */}
      </Fragment>
    </div>
  );
}

export default App;