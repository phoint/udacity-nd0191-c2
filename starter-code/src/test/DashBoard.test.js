import React, { act } from "react";
import { renderWithProviders } from "./testUtils";
import { screen, fireEvent } from "@testing-library/react";
import App from "../component/App";
import { setupStore } from "../app/store";
import { authenticate } from "../features/authedUser/authedUserSlice";
import { fetchQuestions } from "../features/questions/questionSlice";
import { fetchUser } from "../features/users/userSlice";


describe('DashBoard', () => {
    test('will match snapshot', async () => {
        const component = renderWithProviders(<App />, { initialEntries: ['/', '/login'], initialIndex: 0 })
        await act(async () => {
            await component.store.dispatch(fetchUser())
            await component.store.dispatch(fetchQuestions())
            await component.store.dispatch(authenticate({
                email: "sarahedo",
                password: "password123"
            }))
        })
        expect(component).toMatchSnapshot();
    }),

    test('Unauthenticated should be redirect to login page', async () => {
        const component = renderWithProviders(<App />, { initialEntries: ['/', '/login'], initialIndex: 0 })
        await act(async () => {
            await component.store.dispatch(fetchUser())
            await component.store.dispatch(fetchQuestions())
        })
        expect(screen.getByRole('button')).toHaveTextContent('Login');
    })
})