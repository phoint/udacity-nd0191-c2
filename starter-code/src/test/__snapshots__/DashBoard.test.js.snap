// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoard will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h3>
          Dashboard
        </h3>
        <div
          class="dashboard-container"
        >
          <h2>
            New question
          </h2>
          <hr />
          <div
            class="questions"
          >
            <div
              class="question-card"
            >
              <h3>
                mtsamis
              </h3>
              <p>
                02:16:AM | 05/01/2017
              </p>
              <a
                href="/question/xj352vofupe1dqz9emx13r"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
            <div
              class="question-card"
            >
              <h3>
                tylermcginnis
              </h3>
              <p>
                07:09:PM | 03/15/2017
              </p>
              <a
                href="/question/vthrdm985a262al8qx3do"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
          </div>
        </div>
        <div
          class="dashboard-container"
        >
          <h2>
            Done
          </h2>
          <hr />
          <div
            class="questions"
          >
            <div
              class="question-card"
            >
              <h3>
                sarahedo
              </h3>
              <p>
                05:22:AM | 03/04/2017
              </p>
              <a
                href="/question/am8ehyc8byjqgar0jgpub9"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
            <div
              class="question-card"
            >
              <h3>
                tylermcginnis
              </h3>
              <p>
                06:42:PM | 12/24/2016
              </p>
              <a
                href="/question/loxhs1bqm25b708cmbf3g"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
            <div
              class="question-card"
            >
              <h3>
                mtsamis
              </h3>
              <p>
                02:02:PM | 07/14/2016
              </p>
              <a
                href="/question/6ni6ok3ym7mf1p33lnez"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
            <div
              class="question-card"
            >
              <h3>
                sarahedo
              </h3>
              <p>
                09:21:AM | 06/29/2016
              </p>
              <a
                href="/question/8xf0y6ziyjabvozdd253nd"
              >
                <button
                  class="show-button"
                >
                  Show
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h3>
        Dashboard
      </h3>
      <div
        class="dashboard-container"
      >
        <h2>
          New question
        </h2>
        <hr />
        <div
          class="questions"
        >
          <div
            class="question-card"
          >
            <h3>
              mtsamis
            </h3>
            <p>
              02:16:AM | 05/01/2017
            </p>
            <a
              href="/question/xj352vofupe1dqz9emx13r"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
          <div
            class="question-card"
          >
            <h3>
              tylermcginnis
            </h3>
            <p>
              07:09:PM | 03/15/2017
            </p>
            <a
              href="/question/vthrdm985a262al8qx3do"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
        </div>
      </div>
      <div
        class="dashboard-container"
      >
        <h2>
          Done
        </h2>
        <hr />
        <div
          class="questions"
        >
          <div
            class="question-card"
          >
            <h3>
              sarahedo
            </h3>
            <p>
              05:22:AM | 03/04/2017
            </p>
            <a
              href="/question/am8ehyc8byjqgar0jgpub9"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
          <div
            class="question-card"
          >
            <h3>
              tylermcginnis
            </h3>
            <p>
              06:42:PM | 12/24/2016
            </p>
            <a
              href="/question/loxhs1bqm25b708cmbf3g"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
          <div
            class="question-card"
          >
            <h3>
              mtsamis
            </h3>
            <p>
              02:02:PM | 07/14/2016
            </p>
            <a
              href="/question/6ni6ok3ym7mf1p33lnez"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
          <div
            class="question-card"
          >
            <h3>
              sarahedo
            </h3>
            <p>
              09:21:AM | 06/29/2016
            </p>
            <a
              href="/question/8xf0y6ziyjabvozdd253nd"
            >
              <button
                class="show-button"
              >
                Show
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
